clang/test/Analysis/element-region-address-space.c

```diff
@@ -1,11 +1,26 @@
 // RUN: %clang_analyze_cc1 -triple amdgcn-unknown-unknown \
-// RUN:   -analyzer-checker=core -verify %s
+// RUN:   -analyzer-checker=core,unix -verify %s
 
 // expected-no-diagnostics
 //
 // By default, pointers are 64-bits.
+#define ADDRESS_SPACE_64BITS __attribute__((address_space(0)))
 #define ADDRESS_SPACE_32BITS __attribute__((address_space(3)))
 
 int test(ADDRESS_SPACE_32BITS int *p, ADDRESS_SPACE_32BITS void *q) {
   return p == q; // no-crash
 }
+
+// Make sure that the cstring checker handles non-default address spaces
+ADDRESS_SPACE_64BITS void *
+memcpy(ADDRESS_SPACE_64BITS void *,
+       ADDRESS_SPACE_32BITS const void *,
+       long unsigned int);
+
+typedef struct {
+  char m[1];
+} k;
+
+void l(ADDRESS_SPACE_32BITS char *p, ADDRESS_SPACE_64BITS k *n) {
+  memcpy(&n->m[0], p, 4);
+}
```

Comment:
```suggestion
ADDRESS_SPACE_64BITS struct {
  char m[16];
} n;

void avoid_cstring_checker_crash(ADDRESS_SPACE_32BITS char *p) {
  memcpy(&n.m[0], p, 4); // no-crash
}
```
My suggestion does four things:
- it slightly simplifies this testcase (there is no need to typedef the struct type);
- it adds `no-crash` which is the conventional way of marking the location where a crash was occurring (note that `no-crash` has no special meaning for the test executor system, it's just a convention);
- it makes the testcase name more meaningful (feel free to choose a different meaningful name if you prefer something else);
- it increases the size of the array within the struct to clearly avoid overflow warnings.

Note that currently the `alpha.unix.cstring.OutOfBounds` checker is in alpha stage so it isn't active even if you enable the whole `unix` group, but seeing the overflow might be confusing for the readers before they realize this.

Reply:
Good idea, applied.